// core/game/add_gamer.js
//
//
//

const uuid = require('core/uuid/create');
const db = require('db');

function add_gamer(game_uuid, gamer_uuid, cb) {

  cb = cb || function() {};

  if (!game_uuid || !gamer_uuid) {
    return cb({
      status: 400,
      err: new Error('low args')
    });
  }

  db.games.findOne({
    uuid: game_uuid
  }, {
    __v: false
  }).lean().exec(function(err, game) {

    if (err) {

      return cb({
        status: 500,
        err: err
      });
    }

    // if game was full
    if (!game || !game.users_uuid || game.users_uuid.length >= 2) {
      return cb({
        status: 400,
        err: new Error('This game is full of gamer!')
      });
    }

    add_user_to_db(game_uuid, gamer_uuid, cb);
  })

}

module.exports = add_gamer;

///

function add_user_to_db(game_uuid, gamer_uuid, cb) {

  cb = cb || function() {};

  db.games.update({
    uuid: game_uuid
  }, {
    $addToSet: {
      users_uuid: gamer_uuid
    }
  }, function(err, result) {

    if (err) {
      return cb({
        status: 500,
        err: err
      })
    }

    if (result && result.ok && result.n == 1 && result.nModified == 1) {

      cb(null, {
        add: true
      });
    } else if (result && result.ok && result.n == 1) {

      cb(null, {
        exists: true
      });
    }
  })
}
