// core/game/choose.js
//
//
//

const db = require('db');
const Game = require('core/game');

function choose(game_uuid, gamer_uuid, choose, cb) {

  cb = cb || function() {};

  if (!game_uuid || typeof choose == 'undefined') {
    return cb({
      status: 400,
      err: new Error('low args')
    });
  }

  // TODO can player play now?

  db.games.findOne({
    uuid: game_uuid
  }).lean().exec(function(err, game) {

    if (err) {
      return cb({
        status: 500,
        err: err
      })
    }

    let this_game = global.game_schemas[game.game_schema_id];

    this_game.turn(
      choose, game.map,
      game.data || {},
      game.logs || {},
      game.fighters,
      game.turn_user,
      function(err, map, end, data) {

        if (err) {
          // temp act
          return console.log(err);
        }

        // update db
        update_db(game_uuid, map, choose, game.fighters[game.turn_user], data, function(err) {
          if (err) {
            return cb(err);
          }

          // end the game
          if (end) {
            return cb(null, true, game.fighters[game.turn_user]);
          }

          cb(null, false);
        });
      });
  })
}

module.exports = choose;

//////////

function update_db(game_uuid, map, choose, user, data, cb) {

  cb = cb || function() {};

  db.games.update({
    uuid: game_uuid
  }, {
    $set: {
      map: map,
      data: data
    },
    $push: {
      logs: {
        user,
        choose
      }
    }
  }, function(err, result) {

    if (err) {
      return cb({
        status: 500,
        err: err
      })
    }

    return cb(null, true);
  })
}
