// core/game/choose_player.js
//
//
//

const db = require('db');

function choose(game_uuid, cb) {

  cb = cb || function() {};

  let query = {
    uuid: game_uuid
  }

  db.games.findOne(query, {
    fighters: true,
    turn_user: true
  }).lean().exec(function(err, game) {

    if (err) {
      return cb({
        status: 500,
        err: err
      })
    }

    if (!game || !game.fighters || !game.fighters.length) {
      return cb({
        status: 404,
        err: new Error('game or gamers not found!')
      })
    }

    let turn;

    // if we are in middle of the game go to next gamer
    if (typeof game.turn_user != 'undefined') {
      turn = (game.turn_user + 1) % game.fighters.length;
    } else {
      // random choose first fighter
      turn = Math.floor(Math.random() * game.fighters.length);
    }

    db.games.update(query, {
      $set: {
        turn_user: turn
      }
    }, function(err) {

      if (err) {
        return cb({
          status: 500,
          err: err
        })
      }

      return cb(null, true);
    })
  })
}

module.exports = choose;
