// core/game/choose.js
//
//
//

const db = require('db');

function choose(game_uuid, gamer_uuid, choose, cb) {

  cb = cb || function() {};

  if (!game_uuid || typeof choose == 'undefined') {
    return cb({
      status: 400,
      err: new Error('low args')
    });
  }

  // TODO can player play now?

  db.games.findOne({
    uuid: game_uuid
  }).lean().exec(function(err, game) {

    if (err) {
      return cb({
        status: 500,
        err: err
      })
    }

    let this_game = global.game_schemas[game.game_schema_id];

    this_game.turn(
      choose, game.map,
      game.data || {},
      game.logs || {},
      game.fighters[game.turn_user],
      function(err, map, end, data) {

        if (err) {
          // temp act
          return console.log(err);
        }

        if(end){
          console.log("game ends!");
          return;
          // end the game

          cb(null,true);
        }



        // update db
        update_db(game_uuid, map, data);

        // calback for socket call
      });

    cb(null, false);
  })
}

module.exports = choose;

//////////

// function can_i_play_now(game_uuid, gamer_uuid, cb) {
//
//   cb = cb || function() {};
// }


function update_db(game_uuid, map, data, cb) {

  cb = cb || function() {};

  db.games.update({
    uuid: game_uuid
  }, {
    $set: {
      map: map,
      data: data
    }
  }, function(err, result) {

    console.log("updating map in DB...");
    // console.open(result);
  })
}
