// core/game/choose_first_player.js
//
//
//

const db = require('db');

function choose(game_uuid, cb) {

  cb = cb || function() {};

  let query = {
    uuid: game_uuid
  }

  db.games.findOne(query, {
    fighters: true,
    turn_user: true
  }).lean().exec(function(err, game) {

    if (err) {
      return cb({
        status: 500,
        err: err
      })
    }

    if (!game || !game.fighters || !game.fighters.length) {
      return cb({
        status: 404,
        err: new Error('game or gamers not found!')
      })
    }

    // FIXME game can start again if turn_user == 0 and its a BUG
    if (game.turn_user) {
      return cb({
        status: 400,
        err: new Error('game is started and can\'t start again!')
      })
    }

    // random choose first fighter
    let turn = Math.floor(Math.random() * game.fighters.length);

    db.games.update(query, {
      $set: {
        turn_user: turn
      }
    }, function(err) {

      if (err) {
        return cb({
          status: 500,
          err: err
        })
      }

      return cb(null, true);
    })
  })
}

module.exports = choose;
