// core/game/init.js
//
//
//

const db = require('db');
const uuid_create = require('core/uuid/create');

function init(game_schema_id, cb) {

  cb = cb || function() {};

  if (!game_schema_id) {
    return cb({
      status: 400,
      err: new Error('low args')
    })
  }

  uuid_create(function(err, uuid) {

    // TODO : add owner of the game later

    let game = {
      game_schema_id,
      uuid,
      data: {}
    }

    find_game_schema_path(game_schema_id, function(err, game_path) {
      if (err) {
        return cb(err)
      }

      game.map = require(game_path).init().map;

      new db.games(game).save(function(err, game) {

        if (err) {
          return cb({
            status: 500,
            err: err
          });
        }

        return cb(null, game);
      });
    });

  })
}

module.exports = init;

//////////

function find_game_schema_path(game_schema_id, cb) {

  cb = cb || function() {};

  let query = {
    _id: game_schema_id
  }

  db.games_schemas.findOne(query, {
    path: true
  }).lean().exec(function(err, game_schema) {

    if (err) {
      return cb({
        status: 500,
        err: err
      })
    }

    return cb(null, game_schema.path);
  })
}
