// core/game/start.js
//
//
//

const db = require('db');
const Game = require('core/game');
function start(game_uuid, cb) {

  cb = cb || function() {};

  if (!game_uuid) {
    return cb({
      status: 400,
      err: new Error('low args')
    });
  }

  let query = {
    uuid: game_uuid
  };

  db.games.findOne(query).lean().exec(function(err, game) {
    if (err) {
      return cb({
        status: 500,
        err: err
      });
    }

    if (!game) {
      return cb({
        status: 404,
        err: new Error('game not found')
      })
    }

    if (!game.fighter || game.fighter.length != 2) {
      // FIXME
      // this is when a game fighter are not ok!
    }

    db.games.update({
      uuid: game_uuid
    }, {
      $set: {
        starts: true
      }
    }, function(err, result) {
      if (err) {
        return cb({
          status: 500,
          err: err
        });
      }

      // choose first player
      Game.choose_first_player(game_uuid, function(err) {
        console.open(err.err.message);
      });

      return cb(null, result);
    })
  })
}

module.exports = start;
